### ISSUE TITLE
GUI : Field Core Override

### ISSUE THREAD

**Issue by nicoleconser** [ibm-openbmc/dev#2327]
## SME
**BMC**:  @sunharis (for bios attribute questions), @raviteja-b , @ojayanth (for function questions)
**PHYP**: David Stanton
**RAS**: George Ahrens

### Summary
Field Core Override also controls the number of cores the customer can actually use – but you don’t have to pay IBM to activate additional cores since they were paid for in the base price of the server. FCO is primarily used to reduce the number of activated cores in the system (by persistently deconfiguring aka guarding them), which in turn reduces third-party software licensing costs when the vendor has a per-core cost. (e.g. Oracle). FCO is used in the field by low-end customers (via the GUI), whereas mid-range and high-end customers would request the FCO setting prior to shipment, during the purchase process (not via the GUI). 

### User Story
- As a PE, I need to reduce or increase the number or activated cores in the system, so I can control the number of cores the customer can use.
- As an admin, I need to reduce or increase the number or activated cores in the system,so I can reduce third-party software licensing costs when the vendor has a per-core cost. (e.g. Oracle)

--

### ASMI Screen Capture
![image](https://user-images.githubusercontent.com/1183515/114253792-574dc200-9971-11eb-9c6b-0ec857f2bb41.png)

### InVision Prototype
- Under review with backend ~~https://ibm.invisionapp.com/d/main#/console/15347903/319989107/preview~~ https://ibm.invisionapp.com/share/T8O1RA2HBGF

### Tasks
- [ ] Create design review story in webui-vue repo
- [ ] Send email to community sharing the design review story
- [ ] Post in story in the discord webui channel

### References
- eBMC Feature Item: 4.7
- [Link to Discovery folder](https://ibm.ent.box.com/notes/768289701177)

### UX Flow / Interaction Requirements
- [ ] Create new page FCO (Resource Management > under CoD)
- [ ] Create new store for FCO
- [ ] Add Alert
- [ ] Add Current Config section
   - [ ] Add ability to copy config info
- [ ] Add change Config section
   - [ ] values for configuring core range is from 1 through the **licensed** cores available in current config; help text should be dynamic based on # of cores
   - [ ] Add form validation
- [ ] Add info & error toasts

### cURL Commands and Redfish Response
```
https://$bmc/redfish/v1/Systems/system
...
  "ProcessorSummary": {
    "CoreCount": 24, <---- Total installed cores
    "Count": 2,
    "Status": {
      "Health": "OK",
      "HealthRollup": "OK",
      "State": "Disabled"
    }
  },
...

```

-  number of licensed cores = `MaxNumberOfDevices` in `redfish/v1/LicenseService/Licenses/PermProcs`
-  total installed cores = `CoreCount` 
-  upper bound is technically the min of (`CoreCount`  and `MaxNumberOfDevices` )
-  Number of configured cores = `hb_field_core_override`
- `hb_field_core_override` is the pending value (Value we patch to)
https://github.com/openbmc/pldm/blob/67c4be60a8da254e5bdae955bba499e4d2b806e1/oem/ibm/configurations/bios/integer_attrs.json#L81
- `hb_field_core_override_current` is the current value (displayed in the UI).
- For example, System is booted, hb_field_core_override_current=5, user changes field_core_override to 4 but this won't take effect until the next system reboot so the values look like:
hb_field_core_override_current=5
hb_field_core_override=4

### Notes
- [x] Uses the same license service as capacity on demand, follow up with Gunnar

#### Browser Tests (Chrome, Firefox, Safari (Mac), Edge (Windows))
- [ ] Tested responsive layout on supported viewport sizes
- [ ] Verified functionality works as expected, e.g. success and failure use cases
- [ ] Resolved any JavaScript errors thrown to the console

#### Accessibility Tests
- [ ] Verified markup is semantic
- [ ] No reported issues in IBM Accessibility Checker (Free open source tool) https://www.ibm.com/able/toolkit/tools/

#### Test Hooks
- [ ] Verify need for `data-test-id` hooks with test team

**Comment by derick-montague** [comment]
Updating mile stone. List from Dean Sanner indicates this is an `A.1.212` deliverable.

**Comment by derick-montague** [comment]
Users testing internally is complete. We need to discuss further with the BMC Engineer.

**Comment by derick-montague** [comment]
Placed in code review  so the designers can follow up and verify with engineering to validate the design prototype. Will be ready for FED after.

**Comment by ParishrutB** [comment]
In discussion: Whether desired state is saved somewhere before an IPL
https://ibm-systems-power.slack.com/archives/C0Q6TQP5Z/p1622791063425700

**Comment by derick-montague** [comment]
Verified in design exchange that UI data will be available soon and UI is possible.

**Comment by derick-montague** [comment]
The setting itself is a BIOS attribute. The data that is displayed on the screen is needed in order to complete the layout. There is a suggested work around for FEDs to show all processors, which is to query the inventory and create the data. However, @gtmills believes this will be ready in the Redfish API by the end of August.

Unless this data will not be provided by the API, I do not believe the time spent to implement the workaround is useful since it will then need to be removed. The only reason for doing the work would be to check off the done box.

If for some reason the decision is made to implement the workaround, I suggest documenting the API in a new story to assure this is updated. The workaround will require the browser to do more work than it needs to in regard to API calls and creating the data.

**Comment by gtmills** [comment]
> The setting itself is a BIOS attribute. The data that is displayed on the screen is needed in order to complete the layout. There is a suggested work around for FEDs to show all processors, which is to query the inventory and create the data. However, @gtmills believes this will be ready in the Redfish API by the end of August.

> Unless this data will not be provided by the API, I do not believe the time spent to implement the workaround is useful since it will then need to be removed. The only reason for doing the work would be to check off the done box.

> If for some reason the decision is made to implement the workaround, I suggest documenting the API in a new story to assure this is updated. The workaround will require the browser to do more work than it needs to in regard to API calls and creating the data.

https://github.com/ibm-openbmc/dev/issues/3264 is the story we are talking about here

**Comment by dixsie** [comment]
Still waiting on license core information.

**Comment by dixsie** [comment]
Will push/merge in GHE, but need to add upstream later when hardcoded endpoints are resolved.

**Comment by dixsie** [comment]
Bios attribute:
![Screen Shot 2021-11-11 at 11.50.17 AM.png](https://images.zenhubusercontent.com/5c969c26c2488c52df50b7eb/e59de513-9ef2-46d5-86bb-c261b93375ea)

**Comment by dixsie** [comment]
@gtmills 

1. Total installed cores - what is the redfish path and property? Is this `MaxNumberOfDevices` in `/redfish/v1/LicenseService/Licenses/PermProcs`?
2. number of Licensed cores <--- not sure where this value comes from, does it need to be displayed?
3. number of configured cores <--- is this the bios attribute `hb_field_core_override`? 
4. Where does the upper bound value come from?


Bios attribute info: `redfish/v1/Systems/system/Bios`

```
"Attributes": {
....
"hb_field_core_override": 0, <-- patch update
"hb_field_core_override_current": 0, <--- display in UI
}
```

![Screen Shot 2021-11-11 at 11.50.17 AM.png](https://images.zenhubusercontent.com/5c969c26c2488c52df50b7eb/786772c6-9152-4423-ad84-ecaa3b1d65a9)

**Comment by gtmills** [comment]
> Total installed cores - what is the redfish path and property? For total installed cores, you should use:

```
"ProcessorSummary": {
"CoreCount": 0,
```
I'll investigate why this is 0

>  number of Licensed cores 

MaxNumberOfDevices in /redfish/v1/LicenseService/Licenses/PermProcs

> number of configured cores 

Yes, the bios attribute hb_field_core_override

> Where does the upper bound value come from? 

I don't see it in the API, the upper bound is technically the min of (Total installed cores  and MaxNumberOfDevices ) but not sure we should enforce this

**Comment by dixsie** [comment]
Thanks @gtmills, last question, where do we patch to to enable FCO and where do we patch the value?

**Comment by gtmills** [comment]
> Thanks @gtmills, last question, where do we patch to to enable FCO and where do we patch the value?

Opps, missed that question above. 
You will want to set "hb_field_core_override". "hb_field_core_override" is the pending value. 
https://github.com/openbmc/pldm/blob/67c4be60a8da254e5bdae955bba499e4d2b806e1/oem/ibm/configurations/bios/integer_attrs.json#L81

"hb_field_core_override_current" is the current value. 
For example, System is booted, hb_field_core_override_current=5, user changes field_core_override to 4 but this won't take effect until the next system reboot so the values look like: 
hb_field_core_override_current=5
hb_field_core_override=4

**Comment by mzipse** [comment]
This work is complete. Closing.
